classes/solid/terrain/terrain_border.gd

```diff
@@ -0,0 +1,408 @@
+			var opp_cur_ind = 2
+			var opp_next_ind = 3
+			_add_inbetween_segment(
+				areas,
+				cur_group[opp_cur_ind],
+				next_group[opp_next_ind],
+				top_intersect
+			)
+		
+		# Does the bottom of this area intersect the bottom of the next?
+		var bottom_intersect = Geometry.segment_intersects_segment_2d(
+			cur_group[2], cur_group[3],
+			next_group[2], next_group[3]
+		)
+		if bottom_intersect:
+			# TODO: Same code as above, different segment indices.
+			# Set the colliding segments to meet at the same point.
+			var cur_ind = 2
+			var next_ind = 3
+			cur_group[cur_ind] = bottom_intersect
+			next_group[next_ind] = bottom_intersect
+			
+			# Add an area between these.
+			var opp_cur_ind = 1
+			var opp_next_ind = 0
+			_add_inbetween_segment(
+				areas,
+				cur_group[opp_cur_ind],
+				next_group[opp_next_ind],
+				bottom_intersect
+			)
+	
+	# Draw everything
+	# Mark the left-side area for drawing.
+	top_edges.area_queue.append([true, areas.front()])
+	
+	# Draw all areas.
+	for area in areas:
+		# Turn this area into a polygon node.
+		var poly2d = _create_polygon(area.verts, area.normal, z_order, root.top,
+			0 if area.type == "quad" or (area.clock_dir == -1 and area.type == "trio") 
+			else area.verts.size() - 2)
+		
+		# Add it as a child.
+		add_child(poly2d)
+		
+		# Make a duplicate polygon to hold the shadow texture.
+		# TODO: This may be meant to get clipped to within the polygon's body.
+		if !root.tint:
+			var shade = poly2d.duplicate()
+			shade.texture = root.top_shade
+			add_child(shade)
+	
+	# Mark the right-side area for drawing.
+	top_edges.area_queue.append([false, areas.back()])
+
+
+func _add_inbetween_segment(areas, start: Vector2, end: Vector2, circumcenter: Vector2):
+	# A "circumcenter," for clarity, is when you draw a circle that touches all
+	# points of a polygon, then take the center of that circle.
+	
+	# Initialize verts array with starting point.
+	var verts = [start]
+	
+	# Find angle of the (circumcenter -> start) vector relative to the X axis.
+	var s_unit = circumcenter.direction_to(start)
+	var s_angle = atan2(s_unit.y, s_unit.x)
+	# Likewise for (circumcenter -> end).
+	var e_unit = circumcenter.direction_to(end)
+	var e_angle = atan2(e_unit.y, e_unit.x)
+	
+	# Find angle between start and end.
+	# (Currently, only the sign of this is used, so the division isn't needed.)
+	var delta = (e_angle - s_angle) / 5.0
+	
+	# Circular edges, this doesn't work rn, I'll work on it later
+	
+#    #for if e_angle > s_angle
+#    while angle < e_angle:
+#        angle += delta
+#        var vert = circumcenter + Vector2(cos(angle), sin(angle)) * distance
+#        verts.append(vert)
+#    #for if e_angle < s_angle
+#    angle = s_angle
+#    while angle > e_angle:
+#        angle += delta
+#        var vert = circumcenter + Vector2(cos(angle), sin(angle)) * distance
+#        verts.append(vert)
+	
+	# Add the endpoint.
+	verts.append(end)
+	# Add the circumcenter (to make a triangle?).
```

Comment:
Yep, it creates a triangle. It cannot be a quad as there isn't enough space for it.

Originally I intended this to be an arc, but it didn't give good results (hence why we mess a lot with angles before)